Firestore/core/test/firebase/firestore/immutable/tree_sorted_map_test.cc

```diff
@@ -34,6 +34,99 @@ TEST(TreeSortedMap, EmptySize) {
   EXPECT_EQ(Color::Black, map.root().color());
 }
 
+TEST(TreeSortedMap, EmptyHasEmptyChildren) {
+  IntMap map;
+  IntMap::node_type left = map.root().left();
+  ASSERT_TRUE(left.empty());
+
+  IntMap::node_type right = map.root().right();
+  ASSERT_TRUE(right.empty());
+}
+
+TEST(TreeSortedMap, PropertiesForEmpty) {
+  IntMap empty;
+  EXPECT_TRUE(empty.empty());
+  EXPECT_EQ(0, empty.root().value());
+
+  EXPECT_EQ(Color::Black, empty.root().color());
+  EXPECT_FALSE(empty.root().red());
+}
+
+TEST(TreeSortedMap, PropertiesForNonEmpty) {
+  IntMap empty;
+
+  IntMap non_empty = empty.insert(1, 2);
+  EXPECT_FALSE(non_empty.empty());
+  EXPECT_EQ(1, non_empty.root().key());
+  EXPECT_EQ(2, non_empty.root().value());
+
+  // Root nodes are always black
+  EXPECT_EQ(Color::Black, non_empty.root().color());
+  EXPECT_FALSE(non_empty.root().red());
+  EXPECT_TRUE(non_empty.root().left().empty());
+  EXPECT_TRUE(non_empty.root().right().empty());
+}
+
+TEST(TreeSortedMap, RotatesLeft) {
```

Comment:
Okay, this is tricky to do; I've taken a stab at it.

Reply:
Thanks for doing this. I have one entirely optional suggestion: to avoid repeating the path to each node twice (for two test checks), store a reference in a named variable. The variable can then be named the same as in the diagram, which would make it easier to map the code to the picture. E.g.:
```cpp
// Variable names correspond to names of the nodes on the diagram in llrb_node.h
const auto& r = map.root();
EXPECT_EQ(7, r.key());
EXPECT_EQ(Color::Black, r.color());

const auto& x = map.root().left();
EXPECT_EQ(5, x.key());
EXPECT_EQ(Color::Red, x.color());
```

I'm fine if you push back on this.

Reply:
That only really works for final verification so I'm disinclined to doing this.